packages/abi-coder/src/encoding/coders/v1/VecCoder.ts

```diff
@@ -44,15 +47,18 @@ export class VecCoder<TCoder extends Coder> extends Coder<
     const dataLength = length * this.coder.encodedLength;
     const dataBytes = data.slice(offsetAndLength, offsetAndLength + dataLength);
 
-    if (dataBytes.length !== dataLength) {
+    if (!this.#isOptionVec && dataBytes.length !== dataLength) {
       throw new FuelError(ErrorCode.DECODE_ERROR, `Invalid vec byte data size.`);
     }
 
-    return [
-      chunkByLength(dataBytes, this.coder.encodedLength).map(
-        (chunk) => this.coder.decode(chunk, 0)[0]
-      ),
-      offsetAndLength + dataLength,
-    ];
+    let newOffset = offsetAndLength;
+    const chunks = [];
+    for (let i = 0; i < length; i++) {
+      const [decoded, optionOffset] = this.coder.decode(data, newOffset);
+      chunks.push(decoded);
+      newOffset = optionOffset;
+    }
+
+    return [chunks, newOffset];
```

Comment:
As [above](https://github.com/FuelLabs/fuels-ts/pull/2050/files#r1575885670), we can no longer calculate length as `vec len` multiplied by the expected length, as an option could be `None`.